Linux switch extensions

kvadrage opened the pull request:
---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [x] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [x] Is the subject and message clear and concise?
- [x] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [x] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?

bmr-cymru commented on sos/plugins/networking.py:
```diff
@@ -209,6 +234,25 @@ def setup(self):
+            self.add_cmd_output("devlink sb tc bind show")
+            self.add_cmd_output("devlink region show")
+            self.add_cmd_output("devlink dev param show")
+            for dev in self.get_devlink_devs():
```
All the `add_*()` collection methods in `sos` accept either a string with a single item to collect, or a list of strings with multiple items. Please change this to make a single call to `add_cmd_output()` with each `devlink` command as an element in the list argument.

bmr-cymru commented on sos/plugins/networking.py:
```diff
@@ -68,6 +68,31 @@ def get_netns_devs(self, namespace):
+            match = re.match('\s*name\s+(\w+)', line)
+            if match:
+                out.append(match.group(1))
+        return out
```
Ack.

bmr-cymru commented on sos/plugins/networking.py:
```diff
@@ -209,6 +234,25 @@ def setup(self):
+            for dev in self.get_devlink_devs():
+                self.add_cmd_output("devlink -v resource show %s" % dev)
+                self.add_cmd_output("devlink dpipe header show %s" % dev)
+                devlink_table_file = self.get_cmd_output_now("devlink dpipe table show %s" % dev)
```
Ditto - you can just pass the entire list returned by `get_devlink_devs()` into a single call, using a list comprehension to build the commands from the format string and the device list.

bmr-cymru commented on sos/plugins/networking.py:
```diff
@@ -209,6 +234,25 @@ def setup(self):
+                if devlink_table_file:
+                    for table in self.get_devlink_tables(devlink_table_file):
+                        self.add_cmd_output("devlink dpipe table dump %s name %s" % (dev, table))
+
```
Same here.

bmr-cymru commented on sos/plugins/bird.py:
```diff
@@ -0,0 +1,36 @@
+        self.add_cmd_output("birdc show bfd sessions")
+        self.add_cmd_output("birdc show ospf")
+        self.add_cmd_output("birdc show rip")
+        self.add_cmd_output("birdc show static")
```
Please convert to list.

bmr-cymru replied:
Still please convert to a list... ;-)

bmr-cymru replied:
```python
        self.add_cmd_output([
            "birdc show status",
            ...
        ])
```

pmoravec commented:
Hello @kvadrage , could you please follow the suggestions (and rebase your PR to current upstream as a new conflict has appeared)?

kvadrage commented:
> Hello @kvadrage , could you please follow the suggestions (and rebase your PR to current upstream as a new conflict has appeared)?

Hello! Sorry for long delay. I've just rebased the PR and updated it. Some commits were excluded. 
Kindly waiting for your new feedback.

bmr-cymru commented on sos/plugins/networking.py:
```diff
@@ -155,6 +155,54 @@ def setup(self):
+                    "devlink sb port pool show",
+                    "devlink sb tc bind show",
+                    "devlink -s -v trap show"
+                ])
```
This line is over-indented (the closing ']' should align with the 's' in `self.add_cmd_output()`).

kvadrage replied:
Fixed.

bmr-cymru commented on sos/plugins/networking.py:
```diff
@@ -155,6 +155,54 @@ def setup(self):
+                        dev_tables.append((line_dev,
+                                            line_tables.split()[1]))
+            # Capture per-device output from devlink
+            self.add_cmd_output(["devlink sb occupancy snapshot %s" % dev
```
If you're having to split a call like this the preferred format is:

```python
    self.add_cmd_output([
        "devlink sb occupancy snapshot %s" % dev for dev in devices
    ])
```

That said, I think we can improve formatting and readability further here by using a table of commands:

```python
    devlink_cmds = [
        "devlink sb occupancy snapshot %s",
        "devlink sb occupancy show %s",
        ...
    ]

    self.add_cmd_output([cmd % dev for cmd in devlink_cmds for dev in devices])
```

E.g.:

```python
>>> cmds = ["foo %s", "bar %s", "baz %s", "qux %s"]
>>> devs = ["xuq", "zab", "rab", "oof"]
>>> [(cmd % dev) for cmd in cmds for dev in devs]
['foo xuq', 'foo zab', 'foo rab', 'foo oof', 'bar xuq', 'bar zab', 'bar rab', 'bar oof', 'baz xuq', 'baz zab', 'baz rab', 'baz oof', 'qux xuq', 'qux zab', 'qux rab', 'qux oof']
```

kvadrage replied:
Thank you! I rewrote this part.

bmr-cymru commented on sos/plugins/bird.py:
```diff
@@ -0,0 +1,36 @@
+        self.add_cmd_output("birdc show ospf")
+        self.add_cmd_output("birdc show rip")
+        self.add_cmd_output("birdc show static")
+
```
Please add a vim mode line here:


```
# vim: set et ts=4 sw=4 :
```

kvadrage replied:
Fixed

kvadrage commented:
Hi @bmr-cymru !
Can you please check my last commits?

kvadrage commented:
Hi!

Do you have any additional comments for my last commits?

TurboTurtle commented on sos/plugins/networking.py:
```diff
@@ -154,6 +155,54 @@ def setup(self):
+                    "devlink sb pool show",
+                    "devlink sb port pool show",
+                    "devlink sb tc bind show",
+                    "devlink -s -v trap show"
```
These lines are over-indented. 4-spaces indented from the calling line, E.G.:

```python
self.add_cmd_output([
    "devlink", 
    "devlink dev param ",
    "devlink dev info",
    ...
])
```

TurboTurtle commented on sos/plugins/networking.py:
```diff
@@ -154,6 +155,54 @@ def setup(self):
+                # Collect information about devlink dpipe tables
+                devlink_dev_tables = self.exec_cmd(
+                    "devlink dpipe table show %s" % line_dev
+                    )
```
Since we're calling this command later on on line 194, instead of `exec_cmd` we can use `collect_cmd_output` which will both save the output to the archive and give you the output back directly in the `output` dict key.

In this loop there's nothing I immediately see that would need to change, we'd just need to remove the `devlink dpipe table show` entry from `devlink_cmds` below.

TurboTurtle commented on sos/plugins/networking.py:
```diff
@@ -154,6 +155,54 @@ def setup(self):
+        ])
+
+        # Collect information about devlink devices and dpipe tables
+        devlink_dev = self.exec_cmd("devlink dev")
```
Should we save `devlink dev` output? Is it ever useful on its own? If so, convert `exec_cmd` to `collect_cmd_output` here.

TurboTurtle replied:
And now I see we're calling it separately above. So yeah, here we can convert to `collect_cmd_output` and then remove the `devlink dev` entry from line 160.

TurboTurtle commented:
@kvadrage re-ping. This looks to be in good shape still if you'd like for it to be merged. It'd just need a rebase to current master and then the minor `collect_cmd_output` changes noted above.

bmr-cymru reviewed (commented):
Sorry, I hadn't noticed the review request for this PR. If this is still relevant please rebase and re-request review.
